enable non-lexical lifetimes in the MIR borrow checker

nikomatsakis opened the pull request:
This PR, joint work with @spastorino, fills out the NLL infrastructure and integrates it with the borrow checker. **Don't get too excited:** it includes still a number of hacks (the subtyping code is particularly hacky). However, it *does* kinda' work. =)

The final commit demonstrates this by including a test that -- with both the AST borrowck and MIR borrowck -- reports an error by default. But if you pass `-Znll`, you only get an error from the AST borrowck, demonstrating that the integration succeeds:

```
struct MyStruct {
    field: String
}

fn main() {
    let mut my_struct = MyStruct { field: format!("Hello") };

    let value = &my_struct.field;
    if value.is_empty() {
        my_struct.field.push_str("Hello, world!");
        //~^ ERROR cannot borrow (Ast)
    }
}
```

rust-highfive commented:
r? @pnkfelix

(rust_highfive has picked a reviewer for you, use r? to override)

aturon commented:
Congrats, both of you! Been awesome seeing you power through this.

kennytm commented on src/librustc_mir/transform/nll/subtype.rs:
```diff
@@ -0,0 +1,89 @@
+use super::RegionIndex;
```
Tidy error 😏 

```
[00:03:02] tidy error: /checkout/src/librustc_mir/transform/nll/subtype.rs: incorrect license
[00:03:02] tidy error: /checkout/src/test/mir-opt/nll/region-liveness-drop-no-may-dangle.rs:48: line longer than 100 chars
[00:03:03] some tidy checks failed
```

spastorino replied:
Thanks @kennytm I've already fixed this in my local machine, will push in a while :)

kennytm commented on src/test/mir-opt/nll/region-liveness-drop-no-may-dangle.rs:
```diff
@@ -0,0 +1,49 @@
+
+// END RUST SOURCE
+// START rustc.node12.nll.0.mir
+// | R4: {bb1[3], bb1[4], bb1[5], bb2[0], bb2[1], bb2[2], bb3[0], bb4[0], bb4[1], bb4[2], bb6[0], bb7[0], bb7[1], bb8[0]}
```
This needs `// ignore-tidy-linelength`.

spastorino replied:
Also this.

pnkfelix commented on src/librustc_mir/util/liveness.rs:
```diff
@@ -54,29 +54,72 @@ struct BlockInfo {
 
-            // We let Call defined the result in both the success and unwind cases.
-            // This may not be right.
+            // We let Call defined the result in both the success and
```
typo: should be "we let Call *define* the result"

pnkfelix commented on src/librustc_mir/util/liveness.rs:
```diff
@@ -236,25 +270,40 @@ impl<'tcx> Visitor<'tcx> for DefsUses {
+            // DROP USES
 
             // We consider drops to always be uses of locals.
             // Drop eloboration should be run before this analysis otherwise
```
This note about when drop "eloboration" should run (which came from cbdb18650e383fd49beb442ce02d006d0daf865c) ... it applied to @Zoxc's move analysis, but does it apply to our uses here for NLL? I ask because I am pretty sure we want borrowck to run *before* drop elaboration ...?

nikomatsakis replied:
The comment is really kind of outdated. We do *not* want to run drop elaboration before the check -- and we are ok with drops being considered uses.

pnkfelix commented on src/test/mir-opt/nll/region-liveness-two-disjoint-uses.rs:
```diff
@@ -0,0 +1,49 @@
+
+// END RUST SOURCE
+// START rustc.node12.nll.0.mir
+// | R0: {bb1[1], bb2[0], bb2[1]}
```
would it be better to print the region identifiers here as: `#0r`, `#2r`, `#3r` to make more clear the correspondence with their appearances in the borrows?

spastorino replied:
To be honest, at the beginning I have a hard time figuring this output out, so ... maybe going through this and making it more clear is a good things to do. /cc @nikomatsakis

nikomatsakis replied:
I thought about that. I guess we might as well adopt the existing `_#0r`, even though it's horrible.

pnkfelix commented on src/test/mir-opt/nll/region-subtyping-basic.rs:
```diff
@@ -0,0 +1,47 @@
+
+// END RUST SOURCE
+// START rustc.node12.nll.0.mir
+// | R0: {bb1[1], bb1[2], bb1[3], bb1[4], bb1[5], bb1[6], bb2[0], bb2[1]}
```
hmm, do we not have an explicit notation for where R0 (aka `'_#0r`, right?) arises, somewhere in the MIR dump?

pnkfelix commented on src/librustc_mir/borrow_check.rs:
```diff
@@ -46,93 +47,116 @@ pub fn provide(providers: &mut Providers) {
 pub struct MirBorrowckCtxt<'c, 'b, 'a: 'b+'c, 'gcx: 'a+'tcx, 'tcx: 'a> {
-    tcx: TyCtxt<'a, 'gcx, 'gcx>,
-    mir: &'b Mir<'gcx>,
+    tcx: TyCtxt<'a, 'gcx, 'tcx>,
```
niko can i bribe you with some halloween candy to break out this semi-mechanical lifetime-parameter refactoring (or generalization, if you prefer) into a separate commit? This commt is pretty hard to digest the way it is currently set up...

nikomatsakis replied:
Twix.

nikomatsakis replied:
Done btw.

pnkfelix commented:
This is epic!

🥇

pnkfelix commented:
my questions and nits are all pretty minor.

It might be a good idea to see if @arielb1 wants to take a look too, but he might be on vacation

nikomatsakis commented:
@pnkfelix addressed feedback. I also rebased and cleaned up the history, incorporating a number of changes from @spastorino. Have to figure out those travis failures though.

spastorino commented:
@nikomatsakis some of my commits are lost, related to the tests and fixes we were doing last night. Is that included in the code you've pushed? should I push that again?.

nikomatsakis commented:
@spastorino I don't believe they were lost, rather I merged them into the appropriate places.

nikomatsakis commented:
Those travis failures are curious. I don't see them locally, on mac or linux. :/

nikomatsakis commented:
@bors r=pnkfelix

I'm going to go ahead and land this so as to unblock further work. It ought not to deeply interact with any "enabled by default" code, so it's low risk.

bors commented:
:pushpin: Commit aae3e74 has been approved by `pnkfelix`

<!-- @bors r=pnkfelix aae3e74e70e3b21d00142ca9175e9d806d09a2bf -->

badboy commented:
@spastorino Wuhu! Congrats on pushing this through!

bors commented:
:hourglass: Testing commit aae3e74e70e3b21d00142ca9175e9d806d09a2bf with merge 2be4cc040211a85b17f21e813ff62351ae4de642...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.5188), [status-travis](https://travis-ci.org/rust-lang/rust/builds/295881062?utm_source=github_status&utm_medium=notification)
Approved by: pnkfelix
Pushing 2be4cc040211a85b17f21e813ff62351ae4de642 to master...

mgattozzi commented:
It's finally here! :D Well part of it but still! :D
